**3. Import your analysis libraries, as well as your new customer data set as a dataframe.**

In [43]:
# Import library

import pandas as pd
import numpy as np
import os

In [44]:
# Import data

path = r'/Users/peimei/09-2023 Instacart Basket Analysis'
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)
customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

**4. Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.**

In [45]:
# Rename columns

customers.rename(columns = {'First Name':'first_name', 'Surnam':'surname', 'Gender':'gender', 'STATE':'state','Age':'age','n_dependants':'count_dependants'}, inplace = True)

In [46]:
# Check the output

customers.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   count_dependants fam_status  income  
0                 3    married  165665  
1                 0     single   59285  
2                 2    married   99568  
3                 0     single   42049  
4                 1    married   40374

In [47]:
customers.shape

(206209, 10)

In [48]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   user_id           206209 non-null  int64 
 1   first_name        194950 non-null  object
 2   surname           206209 non-null  object
 3   gender            206209 non-null  object
 4   state             206209 non-null  object
 5   age               206209 non-null  int64 
 6   date_joined       206209 non-null  object
 7   count_dependants  206209 non-null  int64 
 8   fam_status        206209 non-null  object
 9   income            206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [49]:
customers.describe()

user_id            age  count_dependants         income
count  206209.000000  206209.000000     206209.000000  206209.000000
mean   103105.000000      49.501646          1.499823   94632.852548
std     59527.555167      18.480962          1.118433   42473.786988
min         1.000000      18.000000          0.000000   25903.000000
25%     51553.000000      33.000000          0.000000   59874.000000
50%    103105.000000      49.000000          1.000000   93547.000000
75%    154657.000000      66.000000          3.000000  124244.000000
max    206209.000000      81.000000          3.000000  593901.000000

**Check all Ranges** <br>

user_id: 1 - 206209<br>
age: 18 - 81<br>
count_dependants: 0 - 3<br>
income: 25903 - 593901<br>

**No abnormal values found.**



**5. Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.**

In [50]:
# Check for NaN values

customers.isnull().sum()

user_id                 0
first_name          11259
surname                 0
gender                  0
state                   0
age                     0
date_joined             0
count_dependants        0
fam_status              0
income                  0
dtype: int64

In [51]:
# Impute NaN values with "Not Provided"

customers['first_name'].fillna('Not Provided', inplace = True)

In [52]:
# Check the output

customers.isnull().sum()

user_id             0
first_name          0
surname             0
gender              0
state               0
age                 0
date_joined         0
count_dependants    0
fam_status          0
income              0
dtype: int64

In [53]:
# Check for mixed-type data

for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

In [54]:
# Change data type of "first_name" column to string

customers['first_name'] = customers['first_name'].astype('str')

In [55]:
# Check the output

for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)
else:
        print('All good.')

All good.


In [56]:
# Check for duplicates

df_dups = customers[customers.duplicated()]

In [57]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, count_dependants, fam_status, income]
Index: []

**6. Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)**

In [58]:
# Import data

ords_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated_2.pkl'))
ords_prods.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   no_of_days_since_last_order  product_id  add_to_cart_order  reordered  \
0                          NaN         196                  1          0   
1                         15.0         196                  1          1   
2                         21.0         196                  1          1   
3                         29.0         196                  1          1   
4                         28.0         196                  1          1   

  product_name  ...    price_range_loc     busiest day    busiest days  \
0         Soda  ...  Mid-range product  Regularly busy  Regularly busy   
1         Soda  ...  Mid-range product  Regularly busy    Slowest days   
2         Soda  ...  Mid-range product  Regularly busy    Slowest days   
3         Soda  ...  Mid-range product      Least busy    Slowest days   
4         Soda  ...  Mid-range product      Least busy    Slowest days   

  busiest_period_of_day max_order  loyalty_flag price_mean spending_flag  \
0        Average orders        10  New customer   6.367797   Low spender   
1        Average orders        10  New customer   6.367797   Low spender   
2           Most orders        10  New customer   6.367797   Low spender   
3        Average orders        10  New customer   6.367797   Low spender   
4           Most orders        10  New customer   6.367797   Low spender   

   median_days_since_last_order   order_frequency_flag  
0                          20.5  Non-frequent customer  
1                          20.5  Non-frequent customer  
2                          20.5  Non-frequent customer  
3                          20.5  Non-frequent customer  
4                          20.5  Non-frequent customer  

[5 rows x 24 columns]

In [59]:
# Combine data frames

ords_prods_customers = ords_prods.merge(customers, on = ['user_id'])
ords_prods_customers.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   no_of_days_since_last_order  product_id  add_to_cart_order  reordered  \
0                          NaN         196                  1          0   
1                         15.0         196                  1          1   
2                         21.0         196                  1          1   
3                         29.0         196                  1          1   
4                         28.0         196                  1          1   

  product_name  ...   order_frequency_flag  first_name  surname  gender  \
0         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   
1         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   
2         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   
3         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   
4         Soda  ...  Non-frequent customer       Linda   Nguyen  Female   

     state age date_joined count_dependants  fam_status income  
0  Alabama  31   2/17/2019                3     married  40423  
1  Alabama  31   2/17/2019                3     married  40423  
2  Alabama  31   2/17/2019                3     married  40423  
3  Alabama  31   2/17/2019                3     married  40423  
4  Alabama  31   2/17/2019                3     married  40423  

[5 rows x 33 columns]

In [60]:
ords_prods_customers.shape

(32399732, 33)

**7. Ensure your notebook contains logical titles, section headings, and descriptive code comments.**

Done.

**8. Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.**

In [61]:
# Save dataframe as pkl file

ords_prods_customers.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_customers.pkl'))

**9. Save your notebook so that you can send it to your tutor for review after completing part 2.**

Done.